tests/specs/lemmas.k

```diff
@@ -74,17 +74,58 @@ module LEMMAS
 
     rule nthbyteof(N, 0, 1) => N
 
+    rule WS ++ .WordStack => WS
 
+    rule #sizeWordStack(#buf(N, _)  , M) => M +Int N
+    rule #sizeWordStack(BUF1 ++ BUF2, M) => M +Int #sizeWordStack(BUF1) +Int #sizeWordStack(BUF2)
+    rule #sizeWordStack(#drop(N, BUF))   => maxInt(#sizeWordStack(BUF) -Int N, 0)
 
+    rule #take(N, BUF)          => BUF                                              requires N ==Int #sizeWordStack(BUF)
+    rule #take(N, BUF1 ++ BUF2) => #take(N, BUF1)                                   requires N <=Int #sizeWordStack(BUF1)
+    rule #take(N, BUF1 ++ BUF2) => BUF1 ++ #take(N -Int #sizeWordStack(BUF1), BUF2) requires N  >Int #sizeWordStack(BUF1)
 
+    rule #drop(N, BUF)           => .WordStack                               requires #sizeWordStack(BUF) <=Int N
+    rule #drop(N, #drop(M, BUF)) => #drop(N +Int M, BUF)
+    rule #drop(N, BUF1 ++ BUF2)  => #drop(N -Int #sizeWordStack(BUF1), BUF2) requires N >=Int #sizeWordStack(BUF1)
+    rule #drop(N, BUF1 ++ BUF2)  => #drop(N, BUF1) ++ BUF2                   requires N <Int #sizeWordStack(BUF1)
 
+    rule #asWord(#buf(N, BUF)) => BUF
 
+    rule #asWord(W0 : W1 : W2 : W3 : #take(28, BUF)) /Int 26959946667150639794667015087019630673637144422540572481103610249216 => #asWord(W0 : W1 : W2 : W3 : .WordStack)
 
+    rule M [ N := BUF ] => M requires #sizeWordStack(BUF) <=Int 0
 
+    rule M [ N := BUF:WordStack ] [ L <- V ]    => M [ L <- V ] [ N := BUF ]                  requires L  <Int N
+    rule M [ N := BUF           ] [ L <- V ]    => M [ L <- V ] [ N := BUF ]                  requires L >=Int N +Int #sizeWordStack(BUF)
+    rule M [ N := BUF           ] [ N <- V ]    => M [ N <- V ] [ N +Int 1 := #drop(1, BUF) ]
+    rule M [ N := BUF           ] [ N := BUF' ] => M [ N := BUF' ]                            requires #sizeWordStack(BUF) ==Int #sizeWordStack(BUF')
 
+    rule #buf(W, BUF) [ L .. K ] => .ByteArray requires L  >Int W
+    rule #buf(W, BUF) [ L .. K ] => .ByteArray requires K <=Int 0
+    rule #buf(W, BUF) [ 0 .. W ] => #buf(W, BUF)
 
+    rule #range(M, N, K) => .ByteArray requires notBool K >Int 0
 
+    rule #range(M [ N := #padToWidth(W, BUF) ], L, K) => #range(M, L, K)
+      requires L +Int K <=Int N
+        orBool N +Int W  <Int L
 
+    rule #range(M [ N := #padToWidth(W, BUF) ], N, W) => BUF
+
+    rule #range(M [ N := BUF:WordStack ], L, K) => #range(M, L, minInt(K, N -Int L)) ++ #range(M [ N := BUF ], N, K -Int minInt(K, N -Int L))
+      requires K >Int 0
+       andBool L <Int N
+
+    rule #range(M [ N := BUF ], L, K) => BUF [ L -Int N .. minInt(K, #sizeWordStack(BUF) -Int (L -Int N)) ] ++ #range(M, N +Int #sizeWordStack(BUF), K -Int minInt(K, #sizeWordStack(BUF) -Int (L -Int N)))
+      requires K  >Int 0
+       andBool L >=Int N
+       andBool L  <Int N +Int #sizeWordStack(BUF)
+
+    rule #range(M [ N := BUF ], L, K) => #range(M, L, K)
+      requires K  >Int 0
+       andBool L >=Int N +Int #sizeWordStack(BUF)
```

Comment:
ok this seems to make sense, but it is quite tricky especially because of using minInt, so i cannot trust my eyeball-checking at this point.  how about writing test specs for all possible cases, especially including the corner cases as well?

Reply:
I've added tests in `tests/specs/functional/lemmas-spec.k`. Have a look.